### PR TITLE
Adding write-up for monitoring using Open Telemetry

### LINKED ISSUE #857: Create Readme for Observability using Opentelemetry
DoD:
- A Readme.md exists for Observability using OpenTelemetry.

Related to: #722 .

Implemented in #867

### PR DESCRIPTION
**Pull request by sreedhar-guda**
# Type of PR
- Documentation changes

*NOTE*: The write-up seems lengthy because of the inline samples. These can be moved as deemed appropriate.

## Purpose

Initial write-up for monitoring using Open Telemetry. The write up includes sample code for:
- OTEL implementation using Collector and no-Collector (Azure monitor) options
- Sample Python code to implement and generate telemetry
- Sample Schema to use for telemetry
- Naming convention suggestions
- Ways to combine execution data with process data

## Does this introduce a breaking change? If yes, details on what can break
No

## Author pre-publish checklist

- [ ] Added test to prove my fix is effective or new feature works
- [x] No PII in logs
- [x] Made corresponding changes to the documentation

## Validation steps

- ...

## Issues Closed or Referenced

- Closes #857

### CODE REVIEW

**Comment by devlace** [comment]
The content is great however, I don't think the Fabric E2E sample is the best place for this content as it's not really *specific* to the Fabric E2E sample but rather data observability in-general using Open Telemetry. Would love ideas on where to move this -- playbook? Top level doc on the repo? We can then link to this from a more condense article that is specific to Fabric E2E sample and even other samples (ie. Databricks).

File: docs/Observability_using_OpenTelemetry.md [hunk 1]
```diff
@@ -0,0 +1,897 @@
+# Observability in Data Projects Using OpenTelemetry
+
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:2]: Remove the following TODOs to merge the PR:

1. the 1st and 2nd of updating sample code link will be resolved in this PR. #933 
2. the others will **NOT** be included in this PR but be tracked with issues #937 , #936 , #935 , #934

File: docs/Observability_using_OpenTelemetry.md [hunk 2]
```diff
@@ -0,0 +1,897 @@
...
+TO DO:
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:3]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 3]
```diff
@@ -0,0 +1,897 @@
...
+
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:4]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 4]
```diff
@@ -0,0 +1,897 @@
...
+- Update notebook links which are good examples for OTEL implementation.
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:5]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 5]
```diff
@@ -0,0 +1,897 @@
...
+- Replace the sample contents for python scripts (otel_monitor_invoker.py and otel_collector_invoker.py) with links.
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:6]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 6]
```diff
@@ -0,0 +1,897 @@
...
+- Where do we keep the config file related to OTEL
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:7]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 7]
```diff
@@ -0,0 +1,897 @@
...
+- What is the preferred setup for Collector setup (VM + Docker)?
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:8]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 8]
```diff
@@ -0,0 +1,897 @@
...
+- Including the monitoring module as part of the Fabric envs
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:9]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 9]
```diff
@@ -0,0 +1,897 @@
...
+- Code sample section updates - Need to add authentication and extensions - Check [Related feature](https://github.com/Azure-Samples/modern-data-warehouse-dataops/issues/715)
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:10]:
```suggestion
```

File: docs/Observability_using_OpenTelemetry.md [hunk 10]
```diff
@@ -0,0 +1,897 @@
...
+  - [Microsoft Fabric monitoring workspace](https://learn.microsoft.com/fabric/admin/monitoring-workspace)
+  - [Microsoft Fabric monitoring hub](https://learn.microsoft.com/fabric/admin/monitoring-hub)
+- Security monitoring
+  - [User activity tracking](https://learn.microsoft.com/fabric/admin/track-user-activities)
+
+This article focuses on application monitoring (user-generated telemetry).
+
+## Monitoring and observability in data projects
+
+Customers heavily rely on dashboards and metrics to evaluate the **business value** offered by a service or process. Therefore, it is crucial to capture all the necessary information at the appropriate granularity and frequency in order to generate these customer-facing dashboards. These dashboards can be operational-focused, functionality/business-focused, or a combination of both. The monitoring and observability (telemetry) data should enable reports on two main aspects of a data system/process:
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:33]:
```suggestion
Customers heavily rely on dashboards and metrics to evaluate the **business value** offered by a service or process. Therefore, it is crucial to capture all the necessary information at the appropriate granularity and frequency in order to generate these customer-facing dashboards. These dashboards can be operational-focused, functionality/business-focused, or a combination of both.   

The monitoring and observability (telemetry) data should enable reports on two main aspects of a data system/process:
```

File: docs/Observability_using_OpenTelemetry.md [hunk 11]
```diff
@@ -0,0 +1,897 @@
...
+1. *Regular development activity* - Use common OTEL library, from above step, in the code to [capture observability data](https://opentelemetry.io/docs/languages/python/instrumentation/).
+
+   - Understand different ways to [create spans](#create-tracesspans).
+   - The [code sample](#creating-telemetry-events-with-opentelemetry-collector) shows a way to generate the telemetry.
+
+### Create telemetry data
+
+The following explains different types of OTEL spans and ways to create them and the [samples](#sample-files) contains examples of a [common schema](#sample-schema) and combining business data with telemetry data.
+
+#### Understand OpenTelemetry span types
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:124]:
```suggestion
#### Understand Span Types
```

File: docs/Observability_using_OpenTelemetry.md [hunk 12]
```diff
@@ -0,0 +1,897 @@
...
+
+Spans are one type of [signals](https://opentelemetry.io/docs/concepts/signals/) generated by OpenTelemetry. There are primarily [two types of Spans](https://opentelemetry-python.readthedocs.io/en/latest/api/trace.html):
```
Comment (siliang-j-1225) [inline docs/Observability_using_OpenTelemetry.md:126]:
```suggestion
There are primarily [two types of Spans](https://opentelemetry-python.readthedocs.io/en/latest/api/trace.html):
```

File: docs/Observability_using_OpenTelemetry.md [hunk 13]
```diff
@@ -0,0 +1,897 @@
...
+        raise # raise or continue as needed
+    else:
+        root_span.set_status(StatusCode.OK) # set success status on root level in case of success
+        logger.info("Setup process is successful.")
+    finally:
+        logger.info(
+            "\n** Setup process is complete. Check the logs for execution status. End of setup process. **\n\n"
+        )
+
+```
```
Comment (yunishimura0716) [inline docs/Observability_using_OpenTelemetry.md:415]:
```suggestion
[Notebook Sample (00_setup.ipynb)](https://github.com/Azure-Samples/modern-data-warehouse-dataops/blob/feat/e2e-fabric-dataops-sample/e2e_samples/fabric_dataops_sample/src/notebooks/00_setup.ipynb)
```

File: docs/Observability_using_OpenTelemetry.md [hunk 14]
```diff
@@ -0,0 +1,897 @@
...
+        """
+        resource = Resource(attributes=metric_resource_attributes)
+        metrics_exporter = AzureMonitorMetricExporter(connection_string=self.conn_string)
+        metrics_reader = PeriodicExportingMetricReader(metrics_exporter)
+        metrics_provider = MeterProvider(resource=resource, metric_readers=[metrics_reader])
+        metrics.set_meter_provider(metrics_provider)
+        meter = metrics.get_meter_provider().get_meter(name=metric_name, version=metric_version)
+
+        return meter
+```
```
Comment (yunishimura0716) [inline docs/Observability_using_OpenTelemetry.md:334]:
```suggestion
[Sample code (otel_monitor_invoker.py)](https://github.com/Azure-Samples/modern-data-warehouse-dataops/blob/feat/e2e-fabric-dataops-sample/e2e_samples/fabric_dataops_sample/config/fabric_environment/otel_monitor_invoker.py)
```

File: docs/Observability_using_OpenTelemetry.md [hunk 15]
```diff
@@ -0,0 +1,897 @@
...
+
+#### Creating telemetry events with OpenTelemetry Collector
+
+See [Creating traces and spans](#create-tracesspans) for typical flow.
+
+Assuming [the code for Collector option](#sample-python-script---using-opentelemetry-collector) is stored as `otel_collector_invoker.py` here is the sample code to use for generating telemetry.
+
+- The code will be same as [telemetry events generation using SDK](#creating-telemetry-events-with-opentelemetry-sdk-no-collector) after replacing the the two lines as shown below:
```
Comment (yunishimura0716) [inline docs/Observability_using_OpenTelemetry.md:584]:
```suggestion
- The code will be same as [telemetry events generation using SDK](https://github.com/Azure-Samples/modern-data-warehouse-dataops/blob/feat/e2e-fabric-dataops-sample/e2e_samples/fabric_dataops_sample/src/notebooks/00_setup.ipynb) after replacing the two lines as shown below:
```

File: docs/Observability_using_OpenTelemetry.md [hunk 16]
```diff
@@ -0,0 +1,897 @@
...
+    - Exception events: "exceptions"
+
+The sample Collector configuration below uses an Environment file(`env.cfg`) placed on the machine where Collector is setup.
+
+```bash
+####################
+# file_name: env.cfg
+#####################
+# Azure App Insights
+APPINSIGHTS CONNECTION STRING="<<App-insgights-connection-string>>"
```
Comment (yunishimura0716) [inline docs/Observability_using_OpenTelemetry.md:790]: Locally tested and double quote (`"`) will be treated as a part of string, so need to remove
```suggestion
APPINSIGHTS CONNECTION STRING=<<App-insgights-connection-string>>
```

**Pull request closed without merging**